src/pam_google_authenticator.c

```diff
@@ -1493,6 +1513,134 @@ static int check_timebased_code(pam_handle_t *pamh, const char*secret_filename,
   return 1;
 }
 
+/* 
+ * Add a 'config' variable that says we logged in from a particular place
+ * at a particular time.  Only remembers the last 10 logins, which 
+ * are replaced in LRU order.
+ *
+ * Returns 0 on success.
+ */
+int
+update_logindetails(pam_handle_t *pamh, const Params *params, char **buf) {
+  const char *rhost = get_rhost(pamh, params);
+  const time_t now = get_time();
+  time_t oldest = now;
+  int oldest_index = -1;
+  unsigned long when = 0;
+  int  found = 0;
+  char *line;
+  char name[] = "LAST ";
+
+  if (rhost == NULL) {
+    return -1;
+  }
+
+  for (int i = 0; i < 10; i++) {
+    char host[40]; /* Max len of ipv6 address is 8*4 digits plus 7 colons.
+                    * Plus trailing NUL is 40
+                    */
+
+    name[4] = i + '0';
+    line = get_cfg_value(pamh, name, *buf);
+    if (line == &oom) {
+      /* Fatal! */
+      return -1;
+    }
+
+    if (!line) {
+      /* Make first empty line the oldest */
+      if (oldest) {
+        oldest_index = i;
+        oldest = 0;
+      }
+      continue;
+    }
+
+    if (sscanf(line, " %40[0-9a-fA-F:.] %lu ", host, &when) != 2) {
```

Comment:
Should be 39, right? This number does not include the terminating null.

Reply:
Yes you're right.